Add new Trigger, matching-type and small fix for the stt correction

corus87 opened the pull request:
- Add Porcupine trigger
- Add matching-type "not-contain"

- Small fix for the STT-Correction:
   - if the order is not in lower case and the correction is or the opposite, it will not be found
   - if one stt-correction contains more than one word, it will not be found

- Add docs for Porcupine and new matching-type

Sispheor commented:
You should run tests locally before pushing. 
It must pass locally before trying on Travis

corus87 commented:
What tests you mean? It was working on my side, also on a fresh raspbian installation. It seems only travis had a problem with the path.

Sispheor commented:
I mean, before pushing any code via a PR, run the tests against the project. All tests must pass before submitting a PR.

LaMonF commented:
Many thanks for this PR @corus87 ! 
#483 
#494
I will have a look on it ASAP. 

Quick suggestions : 
- What about extracting porcupine as a community trigger module that you manage and take care of in your repo ?
First, It will avoid carrying many binaries in the Kalliope core. 
Second it will be easier to  maintain and improve in a separate repository.
Interested users would be able to install it using the `kalliope install` command directly from your repo. What do you think ? 

- Implement unit tests for the use cases you introduced so we avoid any regression in the futur.
- `stt-correction contains more than one word, it will not be found`. Does it mean that I will not be able to correct 2 or 3 words ? For example could I still correct "one two" to "1 2".

Sispheor commented on kalliope/core/OrderAnalyser.py:
```diff
@@ -171,6 +187,31 @@ def _get_split_order_without_bracket(order):
         return split_order
 
+    @classmethod
+    def is_not_contain_matching(cls, user_order, signal_order, not_containing_words):
```
was not already present somewhere else in the core code?

corus87 replied:
Hmm I don't think so, this is for the new matching type, to find the not containing words.

Sispheor commented:
Yes, I agree that it would be interesting to have our first community trigger. 

Thanks for your contribution.

corus87 commented:
Yes sure why not, I can make this to community trigger and take care of it.
@LaMonF  If your stt correction contains two words in one input line, for example: 
```
- input: "thirteen street"
  output: "13th street hd"
```
This will not work if we split the order to search for the input in it. 
I used the changes in the order analyzer for a while now and I have many orders with sst correction, didn't had anymore problems after I changed it. 

So should I close this PR and make a new one only for the matching type and the stt fix? 
Also I could add a new input to the [test-stt-correction.yml](https://github.com/kalliope-project/kalliope/blob/master/Tests/files/test-stt-correction.yml) for the test ?

What about the community trigger then? I guess we need a new site on the kalliope website?

bacardi55 commented on docs/settings/triggers/porcupine.md:
```diff
@@ -0,0 +1,54 @@
+# Trigger engine configuration
+triggers:
+  - porcupine:
+      keyword:
```
Just reading the doc and not the code, but would it be better to have a definition like this:
```
triggers:
  - porcupine:
      keyword:
        - porcupine:
            ppn: "trigger/porcupine/porcupine_raspberrypi.ppn"
            sensitivity: 0.4
          - blueberry_raspberrypi:
            ppn: "trigger/porcupine/blueberry_raspberrypi.ppn"
            sensitivity: 0.4
```

Sispheor replied:
weird indents in what you are showing @bacardi55

bacardi55 replied:
lol sorry on my phone.
The idea is simple though, group both option (ppn file + sensitivity) in the same definition. otherwise it seems weird to define ppn in one place and sensitivity in another and force the user to keep the same orders for both^^

Sispheor replied:
yes you are right. If the trigger can process multiple hotword in the same time, your syntax is easier to read.

corus87 replied:
Yes you are right, it would be much easier to handle the settings. I will take care of it, thanks for the suggestion!

Sispheor commented:
I created a dedicated repo [here](https://github.com/kalliope-project/kalliope_trigger_porcupine) for the trigger and gave you all rights on it @corus87 .

corus87 commented:
Thanks @Sispheor  I will close this one and make a new PR for the matching-type and stt-correction.
